{"componentChunkName":"component---src-components-blog-post-js","path":"/figuring-out/","result":{"data":{"markdownRemark":{"html":"<p>It feels bad trying to debug a code for quite some time and then it turns out to be a very small mistake or a <del>bug</del> feature in the language. So, I thought of sharing it with all, which I feel other developers could also make unknowingly.</p>\n<h2>Story on how I came across</h2>\n<p>I was working with a function in a class to get an available port dynamically and use the port to start a server. This piece of code would be executed parallel on multiple nodes and thus calling the class multiple times. To make things look straight forward, here's how the flow of the code looked like.</p>\n<deckgo-highlight-code language=\"python\"  >\n          <code slot=\"code\">import random\n\ndef get_random_port():\n    return random.randint(0, 100)\n\nclass Server:\n    def __init__(self, port=get_random_port()):\n        self.port = port\n        print(self.port)\n\nif __name__ == &quot;__main__&quot;:\n    server_1 = Server()\n    server_2 = Server()\n    server_3 = Server()</code>\n        </deckgo-highlight-code>\n<p>Everything looks correct, right? Well, not really. The server was not starting saying, \"port already in use\". To debug, I printed the value of the port and it was SAME every time!</p>\n<p>I spent hours debugging the <code>get_random_port()</code> function, but everything looked fine. Even more, the <code>get_random_port()</code> function was working fine at other places. I even removed the parallelization thinking I was doing parallelization wrong, but the same problem was happening even on running it sequentially. I had lost hope and reached out to others asking for help, and turned out it was a very common mistake!</p>\n<h2>What was wrong?</h2>\n<p>It turned out assigning the argument value in the function definition was the issue. Quoting from the <a href=\"https://docs.python.org/3/reference/compound_stmts.html#function-definitions\" target=\"_blank\" rel=\"nofollow\">python function documentation</a>:</p>\n<blockquote>\n<p>Default parameter values are evaluated from left to right when the function definition is executed. This means that the expression is evaluated once when the function is defined and that the same “pre-computed” value is used for each call.</p>\n</blockquote>\n<p>This means the value is evaluated only once when the function is <strong>defined</strong> and not when it is <strong>called.</strong> So the value the port argument got once from <code>get_random_port()</code>, the same was used each time and thus the error was being thrown on starting the server.</p>\n<h2>How to fix it?</h2>\n<p>However, big the problem might sound, the solution to it is very easy and small. Use <code>None</code> as a default value to the argument and assign the value to the argument inside the function. That's it!</p>\n<deckgo-highlight-code language=\"python\"  >\n          <code slot=\"code\">import random\n\ndef get_random_port():\n    return random.randint(0, 100)\n\nclass Server:\n    def __init__(self, port=None):\n        if port is None:\n            self.port = get_random_port()\n            print(self.port)\n\nif __name__ == &quot;__main__&quot;:\n    server_1 = Server()\n    server_2 = Server()\n    server_3 = Server()</code>\n        </deckgo-highlight-code>\n<p>Hope this helps you, and you won't have to spend a lot of time debugging your code. You should NEVER default initialize the argument in a function, unless, you know what you're doing.</p>\n<h2>Side Notes</h2>\n<p>This is also related to <a href=\"https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments\" target=\"_blank\" rel=\"nofollow\">Mutable Default Arguments</a>. It's not a bad design choice for Python, it is helpful for caching and binding a local variable to the current value of an outer variable in a callback. <a href=\"https://stackoverflow.com/questions/9158294/good-uses-for-mutable-function-argument-default-values\" target=\"_blank\" rel=\"nofollow\">Read this StackOverflow thread</a> for some good uses of this.</p>","frontmatter":{"title":"Figuring out","date":"May 28, 2023","description":"Figuring out","keywords":"python, python3, default arguments, bugfix, mutuable default arguments, common mistakes","blog":true}}},"pageContext":{"slug":"/figuring-out"}},"staticQueryHashes":["3172349477","764694655"],"slicesMap":{}}